templates/install_puppet.ps1.erb

```diff
@@ -0,0 +1,185 @@
+# upgrades because the file would lock and cause network stack restarts.
+# See https://tickets.puppetlabs.com/browse/MODULES-4207
+<#
+.Synopsis
+  Fetch the location of the puppet installation from the registry
+#>
+function Script:fetch_install_dir {
+  param()
+  begin {
+    if (Get-ItemProperty -Path "HKLM:\SOFTWARE\Puppet Labs\Puppet" -Name RememberedInstallDir64 -ErrorAction SilentlyContinue) {
+      return (Get-ItemProperty -Path "HKLM:\SOFTWARE\Puppet Labs\Puppet").RememberedInstallDir64
+    } elseif (Get-ItemProperty -Path "HKLM:\SOFTWARE\Puppet Labs\Puppet" -Name RememberedInstallDir -ErrorAction SilentlyContinue) {
+      return (Get-ItemProperty -Path "HKLM:\SOFTWARE\Puppet Labs\Puppet").RememberedInstallDir
+    } else {
+      return $null
+    }
+  }
+}
+
+<#
+.Synopsis
+  Move/rename puppetres.dll to a temporary location
+#>
+function Script:move_puppetres_dll {
+  param()
+  begin {
+    $rand_string = [String[]](Get-Random)
+    $temp_puppetres = "$env:temp\puppetres$rand_string.dll"
+    # _Never_ use the $InstallDir top-level parameter to try and find the InstallDir
+    # for the puppetres workaround. The workaround should _always_ fetch the InstallDir
+    # from the registry. This is so a user can specify a different InstallDir than the
+    # directory where the current package is installed and the workaround will still work
+    # for the already installed package
+    $InstallDir = fetch_install_dir
+    if (Test-Path "$InstallDir\puppet\bin\puppetres.dll") {
+      Out-File -InputObject "Moving puppetres.dll to $temp_puppetres" -FilePath $Logfile -Append
+      Move-Item -Path "$InstallDir\puppet\bin\puppetres.dll" -Destination $temp_puppetres
+    }
+    return $temp_puppetres
+  }
+}
+
+<#
+.Synopsis
+  Restore puppetres.dll to the original location. This should only be used when an installation fails
+.Parameter temp_puppetres
+  Location of the temporary puppetres.dll file.
+#>
+function Script:restore_puppetres_dll {
+  param(
+    [Parameter(Mandatory=$true)]
+    [String] $temp_puppetres
+  )
+  begin {
+    # _Never_ use the $InstallDir top-level parameter to try and find the InstallDir
+    # for the puppetres workaround. The workaround should _always_ fetch the InstallDir
+    # from the registry. This is so a user can specify a different InstallDir than the
+    # directory where the current package is installed and the workaround will still work
+    # for the already installed package
+    $InstallDir = fetch_install_dir
+    if (Test-Path $temp_puppetres) {
+      Out-File -InputObject "Restoring puppetres.dll" -FilePath $Logfile -Append
+      Move-Item -Path $temp_puppetres -Destination "$InstallDir\puppet\bin\puppetres.dll"
+    }
+  }
+}
+
+# ************** Execution start **************
+$ErrorActionPreference = "Stop"
+$service_names=@(
+  "puppet",
+  "pxp-agent",
+  "mcollective"
+)
+$msi_arguments = "/qn /norestart /i `"$Source`" /l*vx+ `"$Logfile`" $InstallArgs"

```

Comment:
I think this should move closer to where the installation happens, as it's easy to forget what these arguments look like later down the code.

Reply:
I think the `InstallArgs` should be appended after everything else like it is in the `install_puppet.bat` file. This would ensure that we don't hit some weird "arguments given out of order" thing for people who pass in their own installation options. Maybe it isn't an issue here, but better safe than sorry.

Reply:
It would not be an issue, but I can move it for posterity.